python/ray/air/execution/barrier.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Callable, List, Optional
+
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class Barrier:
+    """Barrier to collect results and process them in bulk.
+
+    A barrier can be used to collect multiple results and process them in bulk once
+    a certain count or a timeout is reached.
+
+    For instance, if ``max_results=N``, the :meth:`on_completion` callback will be
+    invoked once :meth:`arrive` has been called ``N`` times.
+
+    ``max_results`` can be ``None``, in which case an infinite amount of results
+    will be collected. In this case, the :meth:`on_completion` callback
+    will not be invoked on successful task resolution.
+
+    Args:
+        max_results: Maximum number of results to collect before a call to
+            :meth:`wait` resolves or the :meth:`on_completion` callback is invoked.
+            If ``None``, will collect an infinite number of results.
+        complete_on_first_error: If True, will trigger completion once a single error
+            is received.
+
+    """
+
+    def __init__(
+        self, max_results: Optional[int] = None, complete_on_first_error: bool = False
+    ):
+        raise NotImplementedError
+
+    def arrive(self, *data):
+        """Notify barrier that a result successfully arrived.
+
+        This will count against the ``max_results`` limit. The received result
+        will be included in a call to :meth:`get_results`.
+
+        Args:
+            *data: Result data to be cached. Can be obtained via :meth:`get_results`.
+
+        """
+        raise NotImplementedError
+
+    def error(self, *data):
```

Comment:
nit: Should `*data` instead just be `data: Any`? To ensure compatibility with `get_errors` without wrapping/unwrapping.

Reply:
This is to make barrier compatible with a host of different callback-invoking functions that have different numbers of arguments. E.g. `tracked_task.on_error` will call `barrier.error(exception)` while `tracked_actor_task.on_error` will call `barrier.error(tracked_actor, exception)`.

Unwrapping is up to the caller who controls which kind of callbacks invoke the `barrier.error`

Reply:
In this example you mentioned, if the following are called:

```
barrier.error(exception)
barrier.error(tracked_actor, exception)
```

What would `barrier.get_errors()` return?

Reply:
It would return

```
[[exception], [tracked_actor, exception]]
```

but you should probably not mix different caller types in one barrier.

Reply:
So what I meant is more that we don't want to have different Barrier classes for etc TrackedTask and TrackedActorTask.

Reply:
Ah in that case would it make sense for `get_errors` to return a `List[List[Any]]`? I.e. how do we make sure the usage from `barrier.error()` -> `barrier.get_errors()` is intuitive?

Reply:
Ah yeah, not using `List[List[Any]]` was an oversight. Will update the PR

Reply:
I guess it's actually `List[Tuple[Any]]`. Updated